### DISCUSSION TITLE
Dynamic casting based on passed data

### DISCUSSION THREAD

**Discussion by davidkvasnovsky** [spatie/laravel-data#102] [Q&A] [answered]
Hi, thank you for such a nice package!

I need to pass data from Data Object into my Custom Cast:

`RealEstateData.php`
```php
class RealEstateData extends Data
{
    public function __construct(

        #[WithCast(RealEstateTypeCast::class)]
        public readonly ?RealEstateTypeEnum $type,

        #[WithCast(RealEstatePropertiesCast::class, type: $this->type)] // I know this is not allowed
        public ?object $client_properties,
  ) {
    }
}
```

`RealEstateTypeEnum.php`
```php
enum RealEstateTypeEnum: string
{
    case House = 'house';
    case Apartment = 'apartment';
}
```

Then in my `RealEstatePropertiesCast.php` I would like to do something like this:
```php
class RealEstatePropertiesCast implements Cast
{
    public function __construct(
        protected RealEstateTypeEnum $type,
    ) {
    }

    public function cast(DataProperty $property, mixed $value): mixed
    {
        return match ($this->type) {
            RealEstateTypeEnum::House => RealEstateHousePropertiesData::from($value),
            RealEstateTypeEnum::Apartment => RealEstateApartmentPropertiesData::from($value),

            default => throw CannotCastRealEstateProperties::create($this->type->name, $value),
        };
    }
}

```

The biggest issue is that somehow I am not able to pass enum type of current instance of data object into my cast in a "clean way".

Thanks for any suggestions.

**Comment by rubenvanassche** [comment]
Hi @davidkvasnovsky,

At the moment we do not have such a feature, adding it to the `Cast` interface would break all current casts. That being said, I like the idea and we're working on a v2 of the package and I'm thinking about supporting such a feature(https://github.com/spatie/laravel-data/discussions/84).

**Reply by davidkvasnovsky** [reply]
That would be very nice! May I ask you what alternative do you suggest to use for now?

**Comment by rubenvanassche** [comment, marked as answer]
You could skip the cast and use a magical creation method(https://spatie.be/docs/laravel-data/v1/as-a-data-transfer-object/creating-a-data-object#magical-creation), you have access to all the values being passed over there. I think that's the only solution, for now.
